### PR TITLE
Added ConstWriteAmplificationMergePolicy

### PR DESCRIPTION
**Pull request by fulmicoton**
The `ConstWriteAmplificationMergePolicy` has been designed for a use
case where there are a several index partitions with different sizes,
and partitions tend to be searched separately. (e.g. partitioning by tenant.)

In that case, the StableLogMergePolicy would tend to target the same number
of docs for all tenants. Assuming a merge factor of 10 and a target num docs of 10 millions,
The write amplification observed for a small tenant, emitting splits of 1
document would be 7.

These extra merge have the benefit of making less splits, but really we are
over-trading write amplification for read amplification here.

The `ConstWriteAmplificationMergePolicy` is very simple. It targets a number
of merges instead, and stops once this number of merges is reached.

Only splits with the same number of merge operation are merged together,
and for a given merge operation, we build split in a greedy way.
After sorting the splits per creation date, we append splits one after the
other until we either reach `max_merge_factor` or we exceed the
targetted` split_num_docs`.

### CODE REVIEW

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+/// The `ConstWriteAmplificationMergePolicy` has been designed for a use
+/// case where there are a several index partitions with different sizes,
+/// and partitions tend to be searched separately. (e.g. partitioning by tenant.)
+///
+/// In that case, the StableLogMergePolicy would tend to target the same number
+/// of docs for all tenants. Assuming a merge factor of 10 and a target num docs of 10 millions,
+/// The write amplification observed for a small tenant, emitting splits of 1
+/// document would be 7.
+///
+/// These extra merge have the benefit of making less splits, but really we are
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:36]:
```suggestion
/// These extra merges have the benefit of making less splits, but really we are
```

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+/// over-trading write amplification for read amplification here.
+///
+/// The `ConstWriteAmplificationMergePolicy` is very simple. It targets a number
+/// of merges instead, and stops once this number of merges is reached.
+///
+/// Only splits with the same number of merge operation are merged together,
+/// and for a given merge operation, we build split in a greedy way.
+/// After sorting the splits per creation date, we append splits one after the
+/// other until we either reach `max_merge_factor` or we exceed the
+/// targetted` split_num_docs`.
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:46]:
```suggestion
/// targeted` split_num_docs`.
```

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        ConstWriteAmplificationMergePolicy {
+            max_merge_ops: 3,
+            merge_factor: 3,
+            max_merge_factor: 5,
+            split_num_docs_target: 10_000_000,
+        }
+    }
+
+    /// Merge operations within one level.
+    /// This method assumes that the splits are sorted by inverse creation date.
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:78]: nitpick 

```suggestion
    /// Returns a merge operation within one `num_merge_ops` level.
    /// This method assumes that the splits are sorted by reverse creation date and have all the same `num_merge_ops`.
```

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+    #[test]
+    fn test_const_write_amplification_merge_policy_empty() {
+        let mut splits = Vec::new();
+        let merge_policy = ConstWriteAmplificationMergePolicy::for_test();
+        assert!(merge_policy.operations(&mut splits).is_empty());
+    }
+
+    #[test]
+    fn test_const_write_merge_policy_single() {
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:198]: nitpick

```suggestion
    fn test_const_write_merge_policy_single_split() {
```

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+            ..Default::default()
+        }];
+        let merge_policy = ConstWriteAmplificationMergePolicy::for_test();
+        let operations: Vec<MergeOperation> = merge_policy.operations(&mut splits);
+        assert!(operations.is_empty());
+        assert_eq!(splits.len(), 1);
+    }
+
+    #[test]
+    fn test_const_write_merge_policy_simple() {
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:213]: nitpick 

```suggestion
    fn test_const_write_merge_policy_single_operation() {
```

File: quickwit/quickwit-indexing/src/merge_policy/mod.rs [hunk 6]
```diff
@@ -95,6 +96,13 @@ pub trait MergePolicy: Send + Sync + fmt::Debug {
     fn operations(&self, splits: &mut Vec<SplitMetadata>) -> Vec<MergeOperation>;
     /// A mature split is a split that won't undergo a merge operation in the future.
     fn is_mature(&self, split: &SplitMetadata) -> bool;
+    /// Describe a bunch of properties specific to the given merge policy.
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/merge_policy/mod.rs:99]:
```suggestion
    /// Checks a bunch of properties specific to the given merge policy.
```

File: quickwit/quickwit-indexing/src/merge_policy/mod.rs [hunk 7]
```diff
@@ -160,4 +211,197 @@ pub mod tests {
...
+            assert_eq!(
+                num_splits_in_merge + splits.len(), original_num_splits,
+                "Splits should not be lost."
+            );
+
+            // assert!(
+            //     merge_policy.operations(&mut splits).is_empty(),
+            //     "Merge policy are expected to return all available merge operations."
+            // );
+
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/merge_policy/mod.rs:270]:
```suggestion
```

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+    #[test]
+    fn test_const_write_amp_merge_policy_proptest() {
+        let merge_policy = ConstWriteAmplificationMergePolicy::for_test();
+        crate::merge_policy::tests::proptest_merge_policy(&merge_policy);
+    }
+
+    #[tokio::test]
+    async fn test_simulate_const_write_amplification_merge_policy() -> anyhow::Result<()> {
+        let merge_policy = ConstWriteAmplificationMergePolicy::for_test();
+        let vals = vec![1; 1_211]; //< 1_000 splits with a single doc each.
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:305]:
```suggestion
        let vals = vec![1; 1_211]; //< 1_211 splits with a single doc each.
```

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+    }
+
+    #[test]
+    fn test_const_write_amp_merge_policy_proptest() {
+        let merge_policy = ConstWriteAmplificationMergePolicy::for_test();
+        crate::merge_policy::tests::proptest_merge_policy(&merge_policy);
+    }
+
+    #[tokio::test]
+    async fn test_simulate_const_write_amplification_merge_policy() -> anyhow::Result<()> {
```
Comment (fmassot) [inline quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:303]: Maybe in this simple test we can compute manually the final number of splits after applying all the merges and make an assert on this value.

File: quickwit/quickwit-indexing/src/merge_policy/mod.rs (hunk 7 above)
Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/merge_policy/mod.rs:270]: The stable log merge policy does not observe this actually.

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs (hunk 9 above)
Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:304]: tweaked the simulator to return the final number of splits and added an assert

**Pull request merged**